### PR TITLE
[xcov] fix compatibility issues for plugins depending on FastlaneCore::Project

### LINKED ISSUE #18047: [xcov] Could not find option 'skip_package_dependencies_resolution'
### New Regression Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Regression Information

- Breaking version: [e.g. `2.172.0`]
- Last working version: [e.g. `2.171.0`]

### Regression Description

When updating to the new Fastlane version it looks like the xcov step from our Dangerfile breaks:
```bash
fastlane-2.172.0/fastlane_core/lib/fastlane_core/ui/interface.rb:141:in `user_error!':  (Danger::DSLError)
[!] Invalid `Dangerfile` file: Could not find option 'skip_package_dependencies_resolution'
```

I can find the **skip_package_dependencies_resolution** option added to the diff for the 2.171->2.172 release: https://github.com/fastlane/fastlane/compare/2.171.0...master#diff-83dba92a2b22de647bca3cab87244ae93deeaa94b71b325ead2320cc9ed2b77bR359 and other references to skip_package_dependencies_resolution there... could it be a change in this release triggering it? Maybe Danger-xcov is receiving this option that it does not expect. 

@joshdholtz I am only tagging you because I see your name behind danger-xcov too and you may have better insights here thanks to this.

Dangerfile:
```ruby
# Add a CHANGELOG entry for app changes
modifiedOrNewFiles = (git.added_files + git.modified_files)
if !modifiedOrNewFiles.include?("CHANGELOG.md")
	fail("Please include a CHANGELOG.md entry.")
else
	message "Thanks for adding a CHANGELOG.md entry for this PR."
end

junit.parse "PRJ/test_output/report.junit"
junit.report

xcov.report(
	scheme: 'PRJ',
	workspace: 'PRJ/PRJ.xcworkspace',
	# project: 'PRJ/PRJ.xcodeproj',
	minimum_coverage_percentage: 0.0001,
	derived_data_path: 'PRJ/output_dir',
	ignore_file_path: 'xcovignore/.xcovignore',
	# include_targets: 'PRJ',
	# only_project_targets: false,
	# source_directory: 'PRJ/PRJ',
	configuration: 'Debug'
)
```

##### Complete output when running fastlane, including the stack trace and command used

<details>
  <pre>[!] The exception involves the following plugins:
 -  danger-xcov
bundler: failed to load command: danger (JOB/vendor/bundle/ruby/2.6.0/bin/danger)
JOB/vendor/bundle/ruby/2.6.0/gems/fastlane-2.172.0/fastlane_core/lib/fastlane_core/ui/interface.rb:141:in `user_error!':  (Danger::DSLError)
[!] Invalid `Dangerfile` file: Could not find option 'skip_package_dependencies_resolution' in the list of available options: workspace, project, scheme, configuration, source_directory, derived_data_path, xccov_file_direct_path, output_directory, cloned_source_packages_path, html_report, markdown_report, json_report, minimum_coverage_percentage, slack_url, slack_channel, skip_slack, slack_username, slack_message, ignore_file_path, include_test_targets, exclude_targets, include_targets, only_project_targets, disable_coveralls, coveralls_service_name, coveralls_service_job_id, coveralls_repo_token, xcconfig, ideFoundationPath, legacy_support
 #  from Dangerfile:12
 #  -------------------------------------------
 #  
 >  xcov.report(
 #     scheme: 'SCHEME',
 #  -------------------------------------------
	from JOB/vendor/bundle/ruby/2.6.0/gems/fastlane-2.172.0/fastlane_core/lib/fastlane_core/ui/ui.rb:17:in `method_missing'</pre>
</details>

### Environment

<details>
  <pre><details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                       |
| --------------------------- | ------------------------------------------- |
| OS                          | 11.1                                        |
| Ruby                        | 2.6.6                                       |
| Bundler?                    | true                                        |
| Git                         | git version 2.30.0                          |
| Installation Source         | ~/.rbenv/versions/2.6.6/bin/fastlane        |
| Host                        | macOS 11.1 (20C69)                          |
| Ruby Lib Dir                | ~/.rbenv/versions/2.6.6/lib                 |
| OpenSSL Version             | OpenSSL 1.1.1g  21 Apr 2020                 |
| Is contained                | false                                       |
| Is homebrew                 | false                                       |
| Is installed via Fabric.app | false                                       |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/ |
| Xcode Version               | 12.3                                        |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   | en_US.UTF-8 | ✅ |
| LANGUAGE |             |   |

### fastlane files:

**No Fastfile found**

**No Appfile found**

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.172.0 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                           | Version      |
| ----------------------------- | ------------ |
| bundler                       | 2.2.6        |
| rake                          | 13.0.3       |
| CFPropertyList                | 3.0.3        |
| concurrent-ruby               | 1.1.8        |
| i18n                          | 1.8.7        |
| minitest                      | 5.14.3       |
| thread_safe                   | 0.3.6        |
| tzinfo                        | 1.2.9        |
| activesupport                 | 5.2.4.4      |
| public_suffix                 | 4.0.6        |
| addressable                   | 2.7.0        |
| httpclient                    | 2.8.3        |
| json                          | 2.5.1        |
| algoliasearch                 | 1.27.5       |
| artifactory                   | 3.0.15       |
| atomos                        | 0.1.3        |
| aws-eventstream               | 1.1.0        |
| aws-partitions                | 1.418.0      |
| aws-sigv4                     | 1.2.2        |
| jmespath                      | 1.4.0        |
| aws-sdk-core                  | 3.111.2      |
| aws-sdk-kms                   | 1.41.0       |
| aws-sdk-s3                    | 1.87.0       |
| babosa                        | 1.0.4        |
| claide                        | 1.0.3        |
| colored2                      | 3.1.2        |
| cork                          | 0.3.0        |
| nap                           | 1.1.0        |
| open4                         | 1.3.4        |
| claide-plugins                | 0.9.2        |
| fuzzy_match                   | 2.0.4        |
| netrc                         | 0.11.0       |
| ffi                           | 1.14.2       |
| ethon                         | 0.12.0       |
| typhoeus                      | 1.4.0        |
| cocoapods-core                | 1.10.1       |
| cocoapods-deintegrate         | 1.0.4        |
| cocoapods-downloader          | 1.4.0        |
| cocoapods-plugins             | 1.0.0        |
| cocoapods-search              | 1.0.0        |
| cocoapods-trunk               | 1.5.0        |
| cocoapods-try                 | 1.2.0        |
| escape                        | 0.0.4        |
| fourflusher                   | 2.3.1        |
| gh_inspector                  | 1.1.3        |
| molinillo                     | 0.6.6        |
| ruby-macho                    | 1.4.0        |
| nanaimo                       | 0.3.0        |
| xcodeproj                     | 1.19.0       |
| cocoapods                     | 1.10.1       |
| colored                       | 1.2          |
| highline                      | 1.7.10       |
| commander-fastlane            | 4.4.6        |
| faraday-net_http              | 1.0.1        |
| multipart-post                | 2.0.0        |
| ruby2_keywords                | 0.0.4        |
| faraday                       | 1.3.0        |
| faraday-http-cache            | 2.2.0        |
| rchardet                      | 1.8.0        |
| git                           | 1.8.1        |
| rexml                         | 3.2.4        |
| kramdown                      | 2.3.0        |
| kramdown-parser-gfm           | 1.1.0        |
| no_proxy_fix                  | 0.1.2        |
| sawyer                        | 0.8.2        |
| octokit                       | 4.20.0       |
| unicode-display_width         | 1.7.0        |
| terminal-table                | 1.8.0        |
| danger                        | 8.2.2        |
| ox                            | 2.14.1       |
| danger-junit                  | 1.0.2        |
| danger-plugin-api             | 1.0.0        |
| danger-swiftformat            | 0.8.0        |
| thor                          | 0.20.3       |
| danger-swiftlint              | 0.24.5       |
| dotenv                        | 2.7.6        |
| emoji_regex                   | 3.2.1        |
| excon                         | 0.78.1       |
| unf_ext                       | 0.0.7.7      |
| unf                           | 0.1.4        |
| domain_name                   | 0.5.20190701 |
| http-cookie                   | 1.0.3        |
| faraday-cookie_jar            | 0.0.7        |
| faraday_middleware            | 1.0.0        |
| fastimage                     | 2.2.1        |
| jwt                           | 2.2.2        |
| memoist                       | 0.16.2       |
| multi_json                    | 1.15.0       |
| os                            | 1.1.1        |
| signet                        | 0.14.0       |
| googleauth                    | 0.14.0       |
| mini_mime                     | 1.0.2        |
| declarative                   | 0.0.20       |
| declarative-option            | 0.1.0        |
| uber                          | 0.1.0        |
| representable                 | 3.0.4        |
| retriable                     | 3.1.2        |
| google-api-client             | 0.38.0       |
| digest-crc                    | 0.6.3        |
| webrick                       | 1.7.0        |
| google-apis-core              | 0.2.1        |
| google-apis-iamcredentials_v1 | 0.1.0        |
| google-apis-storage_v1        | 0.1.0        |
| google-cloud-env              | 1.4.0        |
| google-cloud-errors           | 1.0.1        |
| google-cloud-core             | 1.5.0        |
| google-cloud-storage          | 1.30.0       |
| mini_magick                   | 4.11.0       |
| plist                         | 3.6.0        |
| rubyzip                       | 2.3.0        |
| security                      | 0.1.3        |
| naturally                     | 2.2.1        |
| simctl                        | 1.6.8        |
| slack-notifier                | 2.3.2        |
| terminal-notifier             | 2.0.0        |
| tty-screen                    | 0.8.1        |
| tty-cursor                    | 0.7.1        |
| tty-spinner                   | 0.9.3        |
| word_wrap                     | 1.0.0        |
| rouge                         | 2.0.7        |
| xcpretty                      | 0.3.0        |
| xcpretty-travis-formatter     | 1.0.1        |
| xcresult                      | 0.2.1        |
| xcov                          | 1.7.5        |
| danger-xcov                   | 0.5.0        |
</details>

*generated on:* **2021-01-25**
</details></pre>
</details>

### PR DESCRIPTION
**Pull request by ainame**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

Closes https://github.com/fastlane/fastlane/issues/18047

Prior to this PR, https://github.com/fastlane/fastlane/pull/17916 got merged but it appears to cause a regression in `xcov` https://github.com/fastlane/fastlane/issues/18047 https://github.com/fastlane-community/xcov/issues/189  

This is due to the fact https://github.com/fastlane/fastlane/pull/17916 changed what `FastlaneCore::Project` requires its consumers to pass.  Now all the consumers need to pass `disable_package_automatic_updates` and `cloned_source_packages_path` by `options` argument. It's not clearly documented but from the use-cases of `FastlaneCore::Project` `options` is `FastlaneCore::Configuration` object in reality whilst `Hash` is used in unit test cases. So when it comes to fetching a value with a key `FastlaneCore::Configuration` can fail if the key is *not* defined within its available options. That's what reported in those issues.

``` 
Could not find option 'skip_package_dependencies_resolution' in the list of available options: workspace, project, scheme, configuration, source_directory, derived_data_path, xccov_file_direct_path, output_directory, cloned_source_packages_path, html_report, markdown_report, json_report, minimum_coverage_percentage, slack_url, slack_channel, skip_slack, slack_username, slack_message, ignore_file_path, include_test_targets, exclude_targets, include_targets, only_project_targets, disable_coveralls, coveralls_service_name, coveralls_service_job_id, coveralls_repo_token, xcconfig, ideFoundationPath, legacy_support
```

It was no problem in `fastlane` gem itself because in https://github.com/fastlane/fastlane/pull/17916 did cover all the modules that depend on `FastlaneCore::Project` by adding new keys to remain working. However, `xocv` or possibly other gems that depend on `FastlaneCore::Project` is now broken with that reason.

I would say this is very fragile. To remove future regressions with similar reasons, I decided to fix `FastlaneCore::Proejct` rather than `xcov` or others one by one.  

### Description

I didn't want to introduce breaking changes to `FastlaneCore::Project` itself because it may cause another bug😅 What I did is to workaround the validation in `FastlaneCore::Project` when fetching values with `.[]` by temporarily converting `options` to `Hash` each time.  This way we still can expect `FastlaneCore::Configuration` to be passed but we can safely fetch any additional value with a new key in future within `FastlaneCore::Project`.
 
### Testing Steps

See CI passes unit tests. I added a test case to replicate the previously failing situation.

### CODE REVIEW

**Comment by Panajev** [comment]
Hello @rogerluan is there an update on this PR? I am not sure if there is a background discussion on this (probably is), just wanted to understand a bit more about what could be preventing these changes from going live if possible. Thanks :).

**Comment by ainame** [comment]
> I am not sure if there is a background discussion on this (probably is)

@Panajev No, there is not actually. Most of the core team members including me are contributing to fastlane by their spare time. So we respect each one's time.

Do you want to use some changes in 2.172.0 but don't want this regression? If not, you can stay in 2.171.0.

**Comment by bpollman** [comment]
Would love to see this get released. Our CI pipeline could be 3-4 minutes faster by leveraging the new `skip_package_dependencies_resolution` in 2.172.0 but are blocked as we use xcov also. 

Anything I can do to help?

**Comment by ainame** [comment]
@bpollman I appreciate if you test this branch on your project or give me any feedback regarding similar issues if you have any🙇 
You can literally start using this change from now on.

```ruby
gem 'fastlane', git: 'https://github.com/fastlane/fastlane.git', branch: 'fix-xcov-regression'
```

**Comment by Panajev** [comment]
> > I am not sure if there is a background discussion on this (probably is)
> 
> 
> @Panajev No, there is not actually. Most of the core team members including me are contributing to fastlane by their spare time. So we respect each one's time.
> 
> Do you want to use some changes in 2.172.0 but don't want this regression? If not, you can stay with 2.171.0. 

I am going to stay with  2.171.0, but I was not taking about you having to work on it full time, hopefully you did not take it that way: I was not implying there is any  lack of consideration for your free time and the free time of other core team members.

I just thought this might have been discussed on slack or something outside of GitHub and I did not want to make assumptions.

**Comment by Panajev** [comment]
> @bpollman I appreciate if you test this branch on your project and then report us back if it's working or give me any feedback regarding similar issues if you have any🙇 
> 
> You can literally start using this change from now on. Update your `Gemfile` like below. (I understand this is not what you really want😓)
> 
> 
> 
> ```ruby
> 
> gem 'fastlane', git: 'https://github.com/fastlane/fastlane.git', branch: 'fix-xcov-regression'
> 
> ```

Will try and report back

**Review by rogerluan — APPROVED** [review]
Interesting solution! Nice work 💪

File: Gemfile.lock [hunk 1]
```diff
@@ -1,7 +1,7 @@
 PATH
   remote: .
   specs:
-    fastlane (2.171.0)
+    fastlane (2.172.0)
```
Comment (ainame) [inline Gemfile.lock:4]: This should be updated to 2.173.0 by "bundle install" when merging this PR😅

Comment (minuscorp) [reply Gemfile.lock:4]: This shouldn't be pushed into PRs doesn't it? It gets updated in the release process afaik

**Review by minuscorp — COMMENTED** [review]
Just a little comment and LGTM

File: Gemfile.lock (hunk 1 above)
Comment (ainame) [reply Gemfile.lock:4]: Hmm I haven't thought when this should get updated🙁 `Gemfile.lock` started to be checked in from this PR https://github.com/fastlane/fastlane/pull/17892 

I didn't intend to have it work like that way but it sounds fair to me!! For now I'll discard this change and then follow it up in another PR 👍 Thank you for your review.

Comment (joshdholtz) [reply Gemfile.lock]: Ahhhh, I was wondering when Gemfile.lcok was starting to get committed! I totally didn’t see this in PR #17892. This has been messing up the release process a little bit but now I totally understand where it started coming from and how to fix 😊

**Review by joshdholtz — APPROVED** [review]
This is a solid change! Thank you so much for doing this ❤️ I will get this out in a release tonight 💪

**Pull request merged**